### PR TITLE
ceph-osd-device-activation:  a systemd service to trigger underlying LVM device activation

### PR DESCRIPTION
**Pull request by deepfire**
## Description of changes

This addresses point #2 of https://github.com/NixOS/nixpkgs/issues/242152, by creating a service to activate the LVM devices underlying the OSDs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/network-filesystems/ceph.nix [hunk 1]
```diff
@@ -56,6 +56,31 @@ let
...
+    {
+      "ceph-osd-device-activation" =
+      {
+        enable = true;
+        description = "Ceph OSD activation service";
+        after = [ "network-online.target" "time-sync.target" "ceph-mon.target" ];
+        wants = [ "network-online.target" "time-sync.target" ];
+        path = with pkgs; [ lvm2 util-linux ];
+        serviceConfig = {
+          Environment = "CLUSTER=${clusterName}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/network-filesystems/ceph.nix:71]:
```suggestion
  osdActivationService."ceph-osd-device-activation" = {
    enable = true;
    description = "Ceph OSD activation service";
    after = [ "network-online.target" "time-sync.target" "ceph-mon.target" ];
    wants = [ "network-online.target" "time-sync.target" ];
    path = with pkgs; [ lvm2 util-linux ];
    serviceConfig = {
      Environment = "CLUSTER=${cfg.global.clusterName}";
```

**Review by SuperSandro2000 — COMMENTED** [review]
other than the formatting cleanup I can't contribute much here.

@astro can you maybe?

**Comment by nh2** [comment]
I cannot review this PR, as I'm not using the NixOS Ceph service; we use our own NixOS service which does not use `ceph-volume` but writes systemd units explicitly (I dumped that as open source somewhere but want to make it available to NixOS in a more cleaned-up way eventually).

**Comment by deepfire** [comment]
> I cannot review this PR, as I'm not using the NixOS Ceph service; we use our own NixOS service which does not use `ceph-volume` but writes systemd units explicitly (I dumped that as open source somewhere but want to make it available to NixOS in a more cleaned-up way eventually).

@nh2, would you mind sharing that in the imperfect state, in the release-early-release-often paradigm, though? : -)

**Comment by nh2** [comment]
@deepfire Our custom NixOS Ceph service modules: https://gist.github.com/nh2/13425a1f18b4c1ce82edb63c10b163c9

**Comment by johanot** [comment]
@flokli

From: https://docs.ceph.com/en/mimic/ceph-volume/lvm/activate/#summary

image

tldr; `ceph-volume` tries to activate the LVMs themselves if they are not activated already, and... if it was only that, you're right, we wouldn't need it. But, activation step 3 and 4 prepares "stuff" in /var which needs to be present before that start of the OSD's will succeed.

`--no-systemd` just tells ceph-volume to skip step 2, which would fail hard on NixOS anyway.

**Comment by deepfire** [comment]
> @deepfire Our custom NixOS Ceph service modules: https://gist.github.com/nh2/13425a1f18b4c1ce82edb63c10b163c9

Looks very interesting, thank you @nh2!

I'd need to get my long-planned playground Ceph cluster with `nixops-libvirtd`, before I get to hack on that, though..

**Comment by deepfire** [comment]
@SuperSandro2000, done, thanks!

@flokli, what are your considerations now, after @johanot 's clarification?

**Comment by flokli** [comment]
I tried to dig deeper into this, but apparently ceph is broken on nixos-unstable and master.

The first point in #242152 also mentions:

> Use template service files, particularly for the OSD daemons.

I'm wondering how this plays together with this approach, just want to make sure we don't put ourselves into one corner where it's harder to navigate out of.

Spelunking around in (a successful) build, there's [ceph-volume-systemd](https://docs.ceph.com/en/quincy/man/8/ceph-volume-systemd/#:~:text=ceph%2Dvolume%2Dsystemd%20is%20a,volume%20for%20activation%20purposes%20only.), which normally seems to get be called by a `ceph-volume@…` template unit (which we don't ship in NixOS yet).

It seems it takes care of all the necessary activation bits mentioned under the hood as well, so moving to template units should would the necessity for the unit file introduced here alltogether.

**Comment by deepfire** [comment]
@flokli, oh, so we have a better startup point -- so that's great news!

I'd like to play with that, once I get a staging setup with `nixops-libvirtd`.

So, I guess, it'd be wise to deprecate this PR in favor of the better future implementation.
